### PR TITLE
Buffered send in AsyncStreams

### PR DESCRIPTION
**Pull request by amitmurthy**
As discussed, this PR removes type BufferedAsyncStream in favor of incorporating a send buffer in all AsyncStream types.

### CODE REVIEW

**Comment by amitmurthy** [comment]
cc : @JeffBezanson

**Comment by amitmurthy** [comment]
@JeffBezanson , if you are OK with this PR, I'll rebase and merge, post which I can work on  https://github.com/JuliaLang/julia/pull/10170

File: base/multi.jl [hunk 1]
```diff
@@ -174,42 +164,22 @@ function flush_gc_msgs(w::Worker)
...
-function release_write_lock(w::Worker)
-    w.w_io.locked_by = Nullable{Task}()
-    notify(w.w_io.acq_lock_cond)
-end
-
 function send_msg_(w::Worker, kind, args, now::Bool)
     #println("Sending msg $kind")
-    release_lock = acquire_write_lock(w)
+    io = w.w_stream
+    release_lock = lock(io)
```
Comment (vtjnash) [inline base/multi.jl]: i feel like `lock` should be a `do-block` function, to ensure that it must get unlocked before any possible return

e.g
```julia
lock(io) do
    stuff
end
```

**Comment by vtjnash** [comment]
+1

**Comment by amitmurthy** [comment]
Thanks for the review. Have incorporated your suggestions. Appveyor is showing an unrelated (I think) segfault.

**Comment by tkelman** [comment]
on 32 bit it's 10045, on 64 bit there's some new failure in cholesky, should probably open a new issue since i've seen it elsewhere

**Comment by amitmurthy** [comment]
OK. I'll merge this in a few days if there are no other concerns.

**Comment by tkelman** [comment]
does this need doc changes or any new tests?

**Comment by amitmurthy** [comment]
This merely moved the optimization of sending large arrays from a specific implementation in `multi.jl` to being a property of all `AsyncStreams`. We have not yet exported the methods for this functionality since some more work w.r.t. extending it to cover all relevant IO is pending.

Doc changes will happen once the methods are exported.

Tests to cover this have been added as part of the existing implementation - no new tests are required.

**Comment by amitmurthy** [comment]
I just rebased this PR and now OSX Travis is failing with https://travis-ci.org/JuliaLang/julia/jobs/52102333 .

Has anyone seen this before?

**Comment by amitmurthy** [comment]
Restarted the OSX Travis job and it went through.

**Comment by amitmurthy** [comment]
FWIW, the displayed error was
```
The command "cp /tmp/julia/lib/julia/sys.ji local.ji && /tmp/julia/bin/julia-debug -J local.ji -e 'true' && rm local.ji" exited with 134.
1.00s$ cd /tmp/julia/share/julia/test && /tmp/julia/bin/julia-debug --check-bounds=yes runtests.jl all && /tmp/julia/bin/julia-debug --check-bounds=yes runtests.jl pkg
Assertion failed: (bp), function jl_deserialize_value, file dump.c, line 947.
/Users/travis/build.sh: line 41: 10485 Abort trap: 6           /tmp/julia/bin/julia-debug --check-bounds=yes runtests.jl all
The command "cd /tmp/julia/share/julia/test && /tmp/julia/bin/julia-debug --check-bounds=yes runtests.jl all && /tmp/julia/bin/julia-debug --check-bounds=yes runtests.jl pkg" exited with 134.
```

**Comment by timholy** [comment]
I got this `Assertion 'bp' failed` error again for https://travis-ci.org/JuliaLang/julia/jobs/55158379.

**Comment by tkelman** [comment]
@timholy note that when you restart a build on travis, the old logs are gone

**Comment by amitmurthy** [comment]
@tkelman , do you ever sleep? 
:-)

**Comment by tkelman** [comment]
yes but not at normal hours :night_with_stars:

**Pull request merged**